stan/math/prim/functor/hcubature.hpp

```diff
@@ -0,0 +1,521 @@
+    f4 += temp;
+  }
+  free(p4);
+  T_return_type f5 = 0.0;
+  T_return_type* p5
+      = reinterpret_cast<T_return_type*>(malloc(n * sizeof(T_return_type)));
+  for (size_t i = 0; i != g.p[3].size(); i++) {
+    for (size_t j = 0; j != n; j++)
+      p5[j] = deltac[j] * g.p[3][i][j];
+
+    for (size_t j = 0; j != n; j++)
+      cc[j] = c[j] + p5[j];
+    T_return_type temp = integrand(cc, pars);
+    f5 += temp;
+  }
+  free(p5);
+  std::vector<T_return_type>().swap(cc);
+  T_return_type I
+      = v
+        * (g.w[0] * f1 + g.w[1] * f2 + g.w[2] * f3 + g.w[3] * f4 + g.w[4] * f5);
+  T_return_type Idash
+      = v * (g.wd[0] * f1 + g.wd[1] * f2 + g.wd[2] * f3 + g.wd[3] * f4);
+  T_return_type E = fabs(I - Idash);
+
+  int kdivide = 0;
+  T_return_type deltaf = E / (std::pow(10, n) * v);
+  for (size_t i = 0; i != n; i++) {
+    T_return_type delta = divdiff[i] - maxdivdiff;
+    if (delta > deltaf) {
+      kdivide = i;
+      maxdivdiff = divdiff[i];
+    } else if ((fabs(delta) <= deltaf) && (deltac[i] > deltac[kdivide])) {
+      kdivide = i;
+    }
+  }
+  out.result = I;
+  out.err = E;
+  out.kdivide = kdivide;
+  free(deltac);
+  free(divdiff);
+  std::vector<T_return_type>().swap(c);
+}
+
+class Box {
+ public:
+  Box(std::vector<double> a, std::vector<double> b, double& I, double& err,
+      int& kdivide)
+      : a(a), b(b), I(I), E(err), kdiv(kdivide) {}
+  bool operator<(const Box& box) const { return E > box.E; }
+  std::vector<double> a;
+  std::vector<double> b;
+  double I;
+  double E;
+  int kdiv;
+};
+
+inline class Box make_box(int n, std::vector<double> a, std::vector<double> b,
+                          one_d out) {
+  std::vector<double> ac(a);
+  std::vector<double> bc(b);
+  Box box(ac, bc, out.result, out.err, out.kdivide);
+  return box;
+}
+
+}  // namespace internal
+
+/**
+ * Compute the n-dimensional integral of the function \f$f\f$ from \f$a\f$ to
+ \f$b\f$ within
+ * specified relative and absolute tolerances or maximum number of evaluations.
+
+ * \f$a\f$ and \f$b\f$ can be finite or infinite and should be given as vectors.
+ *
+ * The prototype for \f$f\f$ is:
+ \verbatim
+   template <typename T_x, typename T_p>
+   stan::return_type_t<T_x, T_p> f(const T_x& x, const T_p& p) {
+   using T_x_ref = stan::ref_type_t<T_x>;
+   T_x_ref x_ref = x;
+   stan::scalar_seq_view<T_x_ref> x_vec(x_ref);
+   my_params* pars = static_cast<my_params*>(p);
+   type_1 var_1 = (pars->par_1);
+   return ;
+   }
+ \endverbatim
+ *
+ * The parameters can be handed over to f as a struct:
+ \verbatim
+  struct my_params {
+  type_1 par_1;
+  type_2 par_2;
+  };
+ \endverbatim
+ *
+ * @tparam F Type of f
+ * @tparam T_pars Type of paramete-struct
+ * @tparam T_n Type of dimension
+ * @tparam T_a Type of a
+ * @tparam T_b Type of b
+ * @tparam T_maxEval Type of maximum number of evaluations
+ * @tparam T_reqAbsError Type of absolute error
+ * @tparam T_reqRelError Type of relative error
```

Comment:
I think having concrete types instead of templates is good practice but I'm not reviewing that other PR so do whatever the reviewer says.

> Who clicks the "Resolve conversation" button?

Whoever thinks the conversation has been resolved. It doesn't really matter if anyone does, and even after resolving people can still comment or unresolve the conversation. The only difference it makes is that resolved conversations are minimized by default. I don't usually bother to click it at all.